OpenCV is built with CMAKE_BUILD_TYPE None

... which can cause `opencv_xxx-NOTFOUND` make errors in depending packages and eventually affect performance.

Like for most packages released into ROS using bloom and its default templates, OpenCV does not set the `CMAKE_BUILD_TYPE` variable explicitly for binary release builds. For Debian and Ubuntu, bloom adds the `-NDEBUG` compiler option via `DEB_CXXFLAGS_MAINT_APPEND` in [debian/rules.em:19](https://github.com/ros-gbp/opencv3-release/blob/debian/kinetic/opencv3/debian/rules.em#L19), but does not override `CMAKE_BUILD_TYPE` in the `dh_auto_configure` step (cf. https://github.com/ros-infrastructure/bloom/issues/327).

This is a problem for the CMake module configuration installed by OpenCV in certain situations starting in version 3.3.1 because of the following snippet that has been added to [cmake/templates/OpenCVConfig.cmake.in:202ff](https://github.com/ros-gbp/opencv3-release/blob/upstream/3.3.1/cmake/templates/OpenCVConfig.cmake.in#L202):
```cmake
  # OpenCV supports Debug and Release only.
  # RelWithDebInfo and MinSizeRel are mapped to Release
  if(TARGET ${__cvcomponent})
      set_target_properties(${__cvcomponent} PROPERTIES
          MAP_IMPORTED_CONFIG_MINSIZEREL "Release"
          MAP_IMPORTED_CONFIG_RELWITHDEBINFO "Release"
      )
  endif()
```
(see also documentation of the MAP_IMPORTED_CONFIG_&lt;CONFIG&gt; target property [here](https://cmake.org/cmake/help/v3.5/prop_tgt/MAP_IMPORTED_CONFIG_CONFIG.html)).

Because `CMAKE_BUILD_TYPE` was not set when OpenCV was built (or actually is set to `None`, see comment below), the only generated configuration is `NONE`:
```cmake
# from /opt/ros/kinetic/share/OpenCV-3.3.1/OpenCVModules-none.cmake
set_property(TARGET opencv_core APPEND PROPERTY IMPORTED_CONFIGURATIONS NONE)
set_target_properties(opencv_core PROPERTIES
  IMPORTED_LOCATION_NONE "${_IMPORT_PREFIX}/lib/libopencv_core3.so.3.3.1"
  IMPORTED_SONAME_NONE "libopencv_core3.so.3.3"
  )
```

Now to the user package. The typical cmake snippet
```cmake
find_package(OpenCV REQUIRED COMPONENTS core imgproc)
include_directories(${OpenCV_INCLUDE_DIRS})
add_library(foo foo.cpp)
target_link_libraries(foo ${OpenCV_LIBRARIES})
```
works fine if `CMAKE_BUILD_TYPE` is set to anything but `MinSizeRel` or `RelWithDebInfo`. In those cases the [MAP_IMPORTED_CONFIG_&lt;CONFIG&gt;](https://cmake.org/cmake/help/v3.5/prop_tgt/MAP_IMPORTED_CONFIG_CONFIG.html) target property is not set, and CMake apparently automatically finds and selects the only available imported config, `NONE`.

Otherwise, if `CMAKE_BUILD_TYPE` has one of the two values where `MAP_IMPORTED_CONFIG_<CONFIG>` is set to `Release`, linking to the library results in build errors like
```cmake
make[2]: *** No rule to make target 'opencv_core-NOTFOUND', needed by '/opt/xxx/devel/lib/libfoo.so'.  Stop.
```
because CMake would only consider the values of the `IMPORTED_LOCATION_RELEASE` or the general `IMPORTED_LOCATION` target properties as valid library locations, but not `IMPORTED_LOCATION_NONE`.

I found the following snippet in OpenCV's main [CMakeLists.txt](https://github.com/ros-gbp/opencv3-release/blob/upstream/3.3.1/CMakeLists.txt#L507) file:
```cmake
# ----------------------------------------------------------------------------
# OpenCV compiler and linker options
# ----------------------------------------------------------------------------
# In case of Makefiles if the user does not setup CMAKE_BUILD_TYPE, assume it's Release:
if(CMAKE_GENERATOR MATCHES "Makefiles|Ninja" AND "${CMAKE_BUILD_TYPE}" STREQUAL "")
  set(CMAKE_BUILD_TYPE Release)
endif()
```
but the ROS build farm set the `CMAKE_BUILD_TYPE` to `None` explicitly for some reason:
```
10:21:40 dh_auto_configure -- -DCMAKE_INSTALL_PREFIX="/opt/ros/kinetic"
10:21:40 	mkdir -p obj-x86_64-linux-gnu
10:21:40 	cd obj-x86_64-linux-gnu
10:21:40 	cmake .. -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_BUILD_TYPE=None -DCMAKE_INSTALL_SYSCONFDIR=/etc -DCMAKE_INSTALL_LOCALSTATEDIR=/var -DCMAKE_INSTALL_PREFIX=/opt/ros/kinetic
```
(from http://build.ros.org/view/Kbin_uX64/job/Kbin_uX64__opencv3__ubuntu_xenial_amd64__binary/31/consoleFull), so the fallback is not effective.

I am not sure about implications of not building OpenCV with cmake's default Release compiler flags regarding performance, which would include `-O3` additionally to `-NDEBUG`. This might even be a more serious problem than not being able to build depending packages in `RelWithDebInfo` mode.

All of the above has been verified with `ros-kinetic-opencv3` version 3.3.1-0xenial-20171106-101742-0 installed from deb on Ubuntu Xenial and with the default cmake version 3.5.1.

Possible solutions (Debian and Ubuntu releases only):
- patch additional `-DCMAKE_BUILD_TYPE=Release` argument for `dh_auto_configure` in [debian/rules.em:29](https://github.com/ros-gbp/opencv3-release/blob/debian/kinetic/opencv3/debian/rules.em#L29), and eventually remove setting `DEB_CXXFLAGS_MAINT_APPEND` above
- apply patch to upstream so that OpenCV also falls back to `Release` if `CMAKE_BUILD_TYPE` is set to `None` or anything different from `Debug` or `Release`

--- a/debian/rules.em
+++ b/debian/rules.em
@@ -25,8 +25,10 @@ override_dh_auto_configure:
 	# In case we're installing to a non-standard location, look for a setup.sh
 	# in the install tree that was dropped by catkin, and source it.  It will
 	# set things like CMAKE_PREFIX_PATH, PKG_CONFIG_PATH, and PYTHONPATH.
+	# OpenCV needs to be built with CMAKE_BUILD_TYPE Release or (empty), but not
+	# with None as set by debhelper.
 	if [ -f "@(InstallationPrefix)/setup.sh" ]; then . "@(InstallationPrefix)/setup.sh"; fi && \
-	dh_auto_configure -- -DCMAKE_INSTALL_PREFIX="@(InstallationPrefix)"
+	dh_auto_configure -- -DCMAKE_INSTALL_PREFIX="@(InstallationPrefix)" -DCMAKE_BUILD_TYPE=Release
 
 override_dh_auto_build:
 	# In case we're installing to a non-standard location, look for a setup.sh

